File: app/schemas.py
from typing import List, Dict, Any, Optional, Generic, TypeVar
from pydantic import BaseModel, Field, validator
from datetime import datetime
import uuid

T = TypeVar('T')

# =============================================================================
# API ENVELOPE
# =============================================================================

class ApiMeta(BaseModel):
    version: int = 1
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    pagination: Optional[Dict[str, Any]] = None

class ApiError(BaseModel):
    code: str
    message: str
    target: Optional[str] = None # Field name or entity ID
    details: Optional[Any] = None

class ApiResponse(BaseModel, Generic[T]):
    data: Optional[T] = None
    meta: ApiMeta = Field(default_factory=ApiMeta)
    errors: Optional[List[ApiError]] = None

# =============================================================================
# CANONICAL ENTITIES
# =============================================================================

class BaseEntity(BaseModel):
    id: uuid.UUID
    version: int = 1
    created_at: datetime
    updated_at: datetime
    last_modified_by: Optional[uuid.UUID] = None

class ProjectBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    technical_uncertainty: Optional[str] = None
    process_of_experimentation: Optional[str] = None
    qualification_status: str = Field(default="pending")

    @validator('qualification_status')
    def validate_status(cls, v):
        allowed = ["qualified", "not_qualified", "pending", "needs_review"]
        if v not in allowed:
            raise ValueError(f"Status must be one of {allowed}")
        return v

class EmployeeBase(BaseModel):
    name: str
    title: Optional[str] = None
    total_wages: float = Field(default=0.0, ge=0)
    qualified_percent: float = Field(default=0.0, ge=0, le=100)
    state: Optional[str] = None

class ContractorBase(BaseModel):
    name: str
    cost: float = Field(default=0.0, ge=0)
    is_qualified: bool = True
    location: str = "US" # US or Foreign

class BudgetBase(BaseModel):
    name: str
    total_amount: float = Field(default=0.0, ge=0)
    category: Optional[str] = None
    fiscal_year: int = 2024

class ExpenseBase(BaseModel):
    description: str
    amount: float = Field(..., ge=0)
    category: str
    vendor_name: Optional[str] = None
    expense_date: datetime

class SavedViewBase(BaseModel):
    name: str
    entity_type: str
    filters: List[Dict[str, Any]] = []
    sort: List[Dict[str, Any]] = []
    grouping: List[Dict[str, Any]] = []
    visible_columns: List[str] = []
    pinned: bool = False
    is_shared: bool = False

# =============================================================================
# BATCH OPERATIONS
# =============================================================================

class BatchUpdateItem(BaseModel):
    id: uuid.UUID
    updates: Dict[str, Any]

class BatchUpdateResponse(BaseModel):
    success_ids: List[uuid.UUID]
    failed_items: List[Dict[str, Any]] # {id, error}

File: frontend/src/lib/schemas.ts
import { z } from 'zod';

// =============================================================================
// API ENVELOPE SCHEMAS
// =============================================================================

export const ApiMetaSchema = z.object({
  version: z.number().default(1),
  timestamp: z.string(),
  pagination: z.object({
    total: z.number(),
    limit: z.number(),
    offset: z.number(),
  }).optional(),
});

export const ApiErrorSchema = z.object({
  code: z.string(),
  message: z.string(),
  target: z.string().optional(),
  details: z.any().optional(),
});

// =============================================================================
// CANONICAL ENTITY SCHEMAS
// =============================================================================

export const BaseEntitySchema = z.object({
  id: z.string().uuid(),
  version: z.number().default(1),
  created_at: z.string(),
  updated_at: z.string(),
  last_modified_by: z.string().uuid().nullable().optional(),
});

export const ProjectSchema = BaseEntitySchema.extend({
  name: z.string().min(1).max(255),
  description: z.string().nullable().optional(),
  technical_uncertainty: z.string().nullable().optional(),
  process_of_experimentation: z.string().nullable().optional(),
  qualification_status: z.enum(["qualified", "not_qualified", "pending", "needs_review"]).default("pending"),
});

export const EmployeeSchema = BaseEntitySchema.extend({
  name: z.string().min(1),
  title: z.string().nullable().optional(),
  total_wages: z.number().nonnegative().default(0),
  qualified_percent: z.number().min(0).max(100).default(0),
  state: z.string().nullable().optional(),
});

export const ContractorSchema = BaseEntitySchema.extend({
  name: z.string().min(1),
  cost: z.number().nonnegative().default(0),
  is_qualified: z.boolean().default(true),
  location: z.string().default("US"),
});

export const ExpenseSchema = BaseEntitySchema.extend({
  description: z.string().min(1),
  amount: z.number().nonnegative(),
  category: z.string(),
  vendor_name: z.string().nullable().optional(),
  expense_date: z.string(),
});

export const SavedViewSchema = BaseEntitySchema.extend({
  name: z.string().min(1),
  entity_type: z.enum(['projects', 'employees', 'contractors', 'tasks', 'expenses']),
  filters: z.array(z.any()).default([]),
  sort: z.array(z.any()).default([]),
  grouping: z.array(z.any()).default([]),
  visible_columns: z.array(z.string()).default([]),
  pinned: z.boolean().default(false),
  is_shared: z.boolean().default(false),
});

// =============================================================================
// INLINE EDIT SCHEMAS
// =============================================================================

export const InlineEditSchema = z.object({
  field: z.string(),
  value: z.any(),
  version: z.number(), // For conflict detection
});

export type Project = z.infer<typeof ProjectSchema>;
export type Employee = z.infer<typeof EmployeeSchema>;
export type Contractor = z.infer<typeof ContractorSchema>;
export type Expense = z.infer<typeof ExpenseSchema>;
export type SavedView = z.infer<typeof SavedViewSchema>;



